### PR TITLE
Improve Dockerfile

### PR DESCRIPTION
**Pull request by IncognitoJam**
### Motivation
Building the docker image is slow, so I wanted to see if I could speed it up at all.

### Solution
In rosetta-builder, separate fetching dependencies and building the go binary steps so that they can be cached if the dependencies haven't changed (`go.mod` and `go.sum`). This about halved the time to build the image for me if the dependencies didn't change, although I think it'll vary based on your network/pc.

Unrelated to build time I also removed `python` and `python-pip` packages from the dogecoind-builder stage since we don't need them if we're just downloading the binaries.

Also added environment variables (similar to the Golang step) so that the Dogecoin Core version, SHA256 checksum and download URL could be more easily changed in the future.

### CODE REVIEW

**Comment by xanimo** [comment]
I'm trying to build final image with alpine at the moment and clean this up but you beat me to it! 🥂

File: Dockerfile [hunk 1]
```diff
@@ -45,29 +45,36 @@ ENV GOLANG_DOWNLOAD_SHA256 9a58494e8da722c3aef248c9227b0e9c528c7318309827780f162
...
 FROM ubuntu:18.04
 
 RUN apt-get update && \
-  apt-get install --no-install-recommends -y libevent-dev libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-program-options-dev libboost-test-dev libboost-thread-dev && \
-  apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+  apt-get install --no-install-recommends -y \
+    libevent-dev libboost-system-dev libboost-filesystem-dev \
+    libboost-chrono-dev libboost-program-options-dev libboost-test-dev libboost-thread-dev && \
+  apt-get clean && \
+  rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
```
Comment (patricklodder) [inline Dockerfile:77]: I think none of these are needed with a static dogecoind?

Comment (IncognitoJam) [reply Dockerfile:77]: You're right, will remove this too

**Comment by IncognitoJam** [comment]
Is it worth upgrading the Golang version that we build with from 1.15 to 1.16?

**Comment by patricklodder** [comment]
> Is it worth upgrading the Golang version that we build with from 1.15 to 1.16?

I'd like it if we'd experiment with that in a separate PR.

**Review by patricklodder — APPROVED** [review]

**Pull request merged**